design/baremetal-operator/early-cleaning.md

```diff
@@ -0,0 +1,125 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Early out-of-band cleaning
+
+This document proposes running driver-specific out-of-band clean
+steps.
+
+## Status
+
+implementable
+
+## Summary
+
+Depending on the selected driver, BMO will run zero or more out-of-band clean
+steps to ensure the node and the BMC are in the right state. Out-of-band clean
+steps are executed via BMC (e.g. using Redfish protocol) without booting the
+ironic agent (IPA).
+
+The first drivers to receive such handling will be `idrac` and
+`idrac-virtualmedia`. For them we will run the `management.known_good_state`
+clean step that resets iDRAC and clears its job queue.
+
+It is important that these actions run before we try to boot anything on the
+target node, be it the inspection ramdisk or a 3rd party installer ISO.
+
+## Motivation
+
+iDRAC will refuse to create a job if there is already a pending job of the same
+type. To make sure we start in a coherent state, we will reset iDRAC before
+doing anything else to it.
+
+### Goals
+
+* Provide a mechanism to run out-of-band clean steps before inspection.
+* Allow a user to opt out of this mechanism on per node basis.
+* Run `known_good_state` step for iDRAC.
+
+### Non-Goals
+
+* Run any in-band (i.e. using a ramdisk) clean steps.
+* Allow a user to provide the steps to run.
+
+## Proposal
+
+## Design Details
+
+All clean steps run asynchronously in the ironic API, even out-of-band steps
+that are reasonable fast (an order of seconds), thus we need a new state
+`Resetting` to represent the new procedure. So instead of
+
+    Registering -> Inspecting [label="!externallyProvisioned && NeedsHardwareInspection()"]
+
+we will have
+
+    Registering -> Resetting [label="!externallyProvisioned && NeedsReset()"]
+    Registering -> Inspecting [label="!externallyProvisioned && !NeedsReset() && NeedsHardwareInspection()"]
+    Resetting -> Inspecting [label="done && NeedsHardwareInspection()"]
+
+`BareMetalHostSpec` is extended with
+
+```go
+// Skip resetting BMC on registration.
+SkipReset bool `json:"skipReset,omitempty"`
+```
+
+### Implementation Details/Notes/Constraints
+
+The `AccessDetails` interface will be extended with a new call:
+
+```go
+ResetSteps() []nodes.CleanStep
```

Comment:
The `pkg/bmc` package doesn't currently depend on gophercloud. Since that package is used outside of the baremetal-operator it would be good to keep it that way. Can we express the clean steps in some other way? For example, do we need to be explicit about which steps are run, or can we return a bool to say whether the driver supports or requires resetting? Or, could the provisioner just always reset the host?

Reply:
Clean steps depend on a driver... The only thing I can suggest is to copy the `CleanStep` definition in `pkg/bmc`. But is it really useful to avoid this dependency? Why would anyone use `pkg/bmc` but not use `gophercloud` directly or indirectly?

Reply:
Tying `pkg/bmc` to gophercloud forces consumers of `pkg/bmc` to use the same version of gophercloud that we do here.

Perhaps we need a new type in a new package to handle the cleaning interface? Maybe that could live inside `pkg/provisioner/ironic`?